src/crimson/common/tri_mutex.h

```diff
@@ -0,0 +1,159 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/future.hh>
+#include <seastar/core/circular_buffer.hh>
+
+/// shared/exclusive mutual exclusion
+///
+/// similar to seastar::shared_mutex, but instead of two kinds of  waiters,
+/// tri_mutex keeps track of three kinds of them:
+/// - readers
+/// - writers
+/// - exclusive users
+/// and unlike shared_mutex, tri_mutex have two kinds of shared users of lock:
+/// - readers
+/// - writers, which are not mutual-exclusive
```

Comment:
Writers are meant to be exclusive, I'd think.

Reply:
I share the same initial confusion – *non-mutual-exclusive writers* are a bit counter-intuitive. How about commenting on that?

Reply:
agreed. i really need to have a better understanding of pipelined write before explaining this. and this is a bigger topic related to cacheable and rmw operations, AFAICT.

i'd prefer deferring this for a future PR.

Reply:
I'd suggest that the fact that this lock design uses reader and writer is entirely and completely independent of the conventional reader/writer usage.  Here, what we mean is that we can pipeline reads, and we can pipeline writes, but we cannot allow a read while writes are in progress or a write while reads are in progress.  Any rmw operation is therefore exclusive.

Reply:
@athanatos thanks Sam. i will incorporate your comment into the code.

Reply:
#37172